File: packages/canyon-ut/src/server/api/routers/coverage.ts
import { z } from "zod";

import { createTRPCRouter, publicProcedure } from "~/server/api/trpc";

export const coverageRouter = createTRPCRouter({
  hello: publicProcedure
    .input(z.object({ text: z.string() }))
    .query(({ input }) => {
      return {
        greeting: `Hello ${input.text}`,
      };
    }),

  create: publicProcedure
    .input(z.object({ name: z.string().min(1) }))
    .mutation(async ({ ctx, input }) => {
      return ctx.db.coverage.create({
        data: {
      //     model Coverage {
      //   id                String   @id @default(cuid())
      //   sha               String   @map("sha")
      //   branch            String
      //   projectID         String   @map("project_id")
      //   // 代码覆盖率
      //   branchesTotal     Int      @map("branches_total")
      //   branchesCovered   Int      @map("branches_covered")
      //   functionsTotal    Int      @map("functions_total")
      //   functionsCovered  Int      @map("functions_covered")
      //   linesTotal        Int      @map("lines_total")
      //   linesCovered      Int      @map("lines_covered")
      //   statementsTotal   Int      @map("statements_total")
      //   statementsCovered Int      @map("statements_covered")
      //   newlinesTotal     Int      @map("newlines_total")
      //   newlinesCovered   Int      @map("newlines_covered")
      //   summary           String //zstd+pb 必要性，提高概览页面查询速度
      //   // 代码覆盖率详情
      //   // 覆盖率实体
      //   hit               String //zstd+pb
      //   // 通用
      //   createdAt         DateTime @default(now()) @map("created_at") @db.Timestamp(3)
      //
      // @@map("ut_coverage")
      // }
      //     id: "1",
          sha: "1",
          branch: "1",
          projectID: "1",
          branchesTotal: 1,
          branchesCovered: 1,
          functionsTotal: 1,
          functionsCovered: 1,
          linesTotal: 1,
          linesCovered: 1,
          statementsTotal: 1,
          statementsCovered: 1,
          newlinesTotal: 1,
          newlinesCovered: 1,
          summary: "1",
          hit: "1",
        },
      });
    }),

  getLatest: publicProcedure.query(async ({ ctx }) => {
    const coverage = await ctx.db.coverage.findFirst({
      orderBy: { createdAt: "desc" },
    });

    return coverage ?? null;
  }),
});

File: packages/backend/models/coverage.go
package models

import (
	"time"
)

// Coverage represents the coverage data structure
type Coverage struct {
	ID                     string    `json:"id" gorm:"primaryKey;column:id"`
	SHA                    string    `json:"sha" gorm:"column:sha"`
	Branch                 string    `json:"branch" gorm:"column:branch"`
	CompareTarget          string    `json:"compareTarget" gorm:"column:compare_target"`
	Provider               string    `json:"provider" gorm:"column:provider"`
	BuildProvider          string    `json:"buildProvider" gorm:"column:build_provider"`
	BuildID                string    `json:"buildId" gorm:"column:build_id"`
	ProjectID              string    `json:"projectId" gorm:"column:project_id"`
	Reporter               string    `json:"reporter" gorm:"column:reporter"`
	ReportID               string    `json:"reportId" gorm:"column:report_id"`
	CovType                string    `json:"covType" gorm:"column:cov_type"`
	BranchesTotal          int       `json:"branchesTotal" gorm:"column:branches_total"`
	BranchesCovered        int       `json:"branchesCovered" gorm:"column:branches_covered"`
	FunctionsTotal         int       `json:"functionsTotal" gorm:"column:functions_total"`
	FunctionsCovered       int       `json:"functionsCovered" gorm:"column:functions_covered"`
	LinesTotal             int       `json:"linesTotal" gorm:"column:lines_total"`
	LinesCovered           int       `json:"linesCovered" gorm:"column:lines_covered"`
	StatementsTotal        int       `json:"statementsTotal" gorm:"column:statements_total"`
	StatementsCovered      int       `json:"statementsCovered" gorm:"column:statements_covered"`
	NewlinesTotal          int       `json:"newlinesTotal" gorm:"column:newlines_total"`
	NewlinesCovered        int       `json:"newlinesCovered" gorm:"column:newlines_covered"`
	ChangebranchesCovered  int       `json:"changebranchesCovered" gorm:"column:changebranches_covered"`
	ChangebranchesTotal    int       `json:"changebranchesTotal" gorm:"column:changebranches_total"`
	ChangefunctionsCovered int       `json:"changefunctionsCovered" gorm:"column:changefunctions_covered"`
	ChangefunctionsTotal   int       `json:"changefunctionsTotal" gorm:"column:changefunctions_total"`
	Summary                []byte    `json:"-" gorm:"column:summary"`
	Hit                    []byte    `json:"-" gorm:"column:hit"`
	InstrumentCwd          string    `json:"instrumentCwd" gorm:"column:instrument_cwd"`
	CreatedAt              time.Time `json:"createdAt" gorm:"column:created_at"`
	UpdatedAt              time.Time `json:"updatedAt" gorm:"column:updated_at"`
}

// TableName specifies the table name for GORM
func (Coverage) TableName() string {
	return "coverage"
}

// CoverageListRequest represents the request parameters for listing coverage
type CoverageListRequest struct {
	Page      int    `form:"page" binding:"min=1"`
	PageSize  int    `form:"pageSize" binding:"min=1,max=100"`
	ProjectID string `form:"projectId"`
	Branch    string `form:"branch"`
	Provider  string `form:"provider"`
	CovType   string `form:"covType"`
}

// CoverageListResponse represents the paginated response
type CoverageListResponse struct {
	Data       []Coverage `json:"data"`
	Total      int64      `json:"total"`
	Page       int        `json:"page"`
	PageSize   int        `json:"pageSize"`
	TotalPages int        `json:"totalPages"`
}

// DefaultPageSize is the default page size if not specified
const DefaultPageSize = 20

// MaxPageSize is the maximum allowed page size
const MaxPageSize = 100